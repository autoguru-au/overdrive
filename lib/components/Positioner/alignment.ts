export enum EAlignment {
	TOP = 'top',
	TOP_LEFT = 'top_left',
	TOP_RIGHT = 'top_right',
	BOTTOM = 'bottom',
	BOTTOM_LEFT = 'bottom_left',
	BOTTOM_RIGHT = 'bottom_right',
	LEFT = 'left',
	RIGHT = 'right',
}
