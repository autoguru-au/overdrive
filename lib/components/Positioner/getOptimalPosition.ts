import { EAlignment } from './alignment';

export interface Dimensions {
	width: number;
	height: number;
}

export interface Position {
	top: number;
	left: number;
}

export interface Rect extends Dimensions, Position {
	right: number;
	bottom: number;
}

export interface AlignmentRect {
	alignment: EAlignment;
	rect: Rect;
}

const DEFAULT_OFFSET = 12;

/**
 * @description
 * The method that will give you the final position, (top, left) values in pixels of where the container needs to be
 *     placed, so that it.
 *     - A: Fit's in the viewport
 *     - B: Places itself at your proffered alignment, if not, it'll be flipped.
 */
export const getOptimalPosition = (
	preferredAlignment: EAlignment,
	triggerRect: Readonly<Rect>, // The rect of the trigger
	containerDimensions: Readonly<Dimensions>, // The width, height of the container
	triggerOffset: number = DEFAULT_OFFSET, // The pixels to leave around the trigger
	viewportOffset: number = DEFAULT_OFFSET, // The pixels to leave around the viewport
	viewportDimensions: Readonly<Dimensions> = getViewportDimensions(),
): AlignmentRect => {
	const { rect, alignment } = getFittedPosition(
		preferredAlignment,
		containerDimensions,
		triggerRect,
		triggerOffset,
		viewportOffset,
		viewportDimensions,
	);

	return {
		alignment,
		rect: handleOverflow(rect, viewportOffset, viewportDimensions),
	};
};

/**
 * @description
 * Given a computed rect, finally we need to handle the overflow viewport "locking" of this rect.
 */
const handleOverflow = (
	rect: Rect,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
): Rect => {
	// Push rect to the right if overflowing on the left side of the viewport.
	if (rect.left < viewportOffset) {
		rect.right += Math.ceil(Math.abs(rect.left - viewportOffset));
		rect.left = Math.ceil(viewportOffset);
	}

	// Push rect to the left if overflowing on the right side of the viewport.
	if (rect.right > viewportDimensions.width - viewportOffset) {
		const delta = Math.ceil(
			rect.right - (viewportDimensions.width - viewportOffset),
		);
		rect.left -= delta;
		rect.right -= delta;
	}

	// Push rect down if overflowing on the top side of the viewport.
	if (rect.top < viewportOffset) {
		rect.top += Math.ceil(Math.abs(rect.top - viewportOffset));
		rect.bottom = Math.ceil(viewportOffset);
	}

	// Push rect up if overflowing on the bottom side of the viewport.
	if (rect.bottom > viewportDimensions.height - viewportOffset) {
		const delta = Math.ceil(
			rect.bottom - (viewportDimensions.height - viewportOffset),
		);
		rect.top -= delta;
		rect.bottom -= delta;
	}

	return rect;
};

/**
 * @description
 * Give this method a preferred alignment - and we'll give you back an alignment and rect for where we think the best
 *     place to place it. By best we mean, if the consumer prefers it on the LEFT, but there is no more room left on
 *     the LEFT, we'll place it to right (we flip).
 *
 * The graph here illustrates the difference between horizontal and vertical, its more about on what axis we flip when
 *     there's no more room.
 *
 *  ---------------- TOP ------------------ <- Vertical
 *        +---------------------+
 *        |                     |
 *  LEFT  |       trigger       |  RIGHT -- <- Horizontal
 *        |                     |
 *        +---------------------+
 *  -------------- BOTTOM ----------------- <- Vertical
 */
const getFittedPosition = (
	preferredAlignment: EAlignment,
	containerDimensions: Readonly<Dimensions>,
	triggerRect: Readonly<Rect>,
	triggerOffset: number,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
): AlignmentRect =>
	(isAlignedHorizontal(preferredAlignment)
		? handleIsHorizontal
		: handleIsVertical)(
		preferredAlignment,
		viewportOffset,
		viewportDimensions,
		getRectForAlignmentFactory(
			triggerRect,
			triggerOffset,
			containerDimensions,
		),
	);

const handleIsHorizontal = (
	preferredAlignment: EAlignment,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
	genRectFn: ReturnType<RectForAlignmentFactoryFn>,
): AlignmentRect => {
	const isFitsOnLeft = (rect: Rect): boolean => rect.left > viewportOffset;
	const isFitsOnRight = (rect: Rect): boolean =>
		rect.right < viewportDimensions.width - viewportOffset;

	const leftRect = genRectFn(EAlignment.LEFT);
	const rightRect = genRectFn(EAlignment.RIGHT);

	const fitsOnLeft = isFitsOnLeft(leftRect);
	const fitsOnRight = isFitsOnRight(rightRect);

	if (preferredAlignment === EAlignment.LEFT) {
		if (fitsOnLeft) {
			return {
				alignment: EAlignment.LEFT,
				rect: leftRect,
			};
		}

		if (fitsOnRight) {
			return {
				alignment: EAlignment.RIGHT,
				rect: rightRect,
			};
		}
	}

	if (preferredAlignment === EAlignment.RIGHT) {
		if (fitsOnRight) {
			return {
				alignment: EAlignment.RIGHT,
				rect: rightRect,
			};
		}

		if (fitsOnLeft) {
			return {
				alignment: EAlignment.LEFT,
				rect: leftRect,
			};
		}
	}

	const spaceRight = Math.abs(
		viewportDimensions.width - viewportOffset - rightRect.right,
	);
	const spaceLeft = Math.abs(leftRect.left - viewportOffset);

	if (spaceRight < spaceLeft) {
		return {
			alignment: EAlignment.RIGHT,
			rect: rightRect,
		};
	}

	return {
		alignment: EAlignment.LEFT,
		rect: leftRect,
	};
};

const handleIsVertical = (
	preferredAlignment: EAlignment,
	viewportOffset: number,
	viewportDimensions: Readonly<Dimensions>,
	genRectFn: ReturnType<RectForAlignmentFactoryFn>,
): AlignmentRect => {
	const isFitsOnTop = (rect: Rect): boolean => rect.top > viewportOffset;
	const isFitsOnBottom = (rect: Rect): boolean =>
		rect.bottom < viewportDimensions.height - viewportOffset;

	const isTop = isAlignedOnTop(preferredAlignment);

	const topRect = genRectFn(
		isTop ? preferredAlignment : flipHorizontal(preferredAlignment),
	);
	const bottomRect = genRectFn(
		isTop ? flipHorizontal(preferredAlignment) : preferredAlignment,
	);

	if (isTop) {
		if (isFitsOnTop(topRect)) {
			return {
				alignment: preferredAlignment,
				rect: topRect,
			};
		}

		if (isFitsOnBottom(bottomRect)) {
			return {
				alignment: flipHorizontal(preferredAlignment),
				rect: bottomRect,
			};
		}
	} else {
		if (isFitsOnBottom(bottomRect)) {
			return {
				alignment: preferredAlignment,
				rect: bottomRect,
			};
		}

		if (isFitsOnTop(topRect)) {
			return {
				alignment: flipHorizontal(preferredAlignment),
				rect: topRect,
			};
		}
	}

	const spaceBottom = Math.abs(
		viewportDimensions.height - viewportOffset - bottomRect.bottom,
	);
	const spaceTop = Math.abs(topRect.top - viewportOffset);

	return spaceBottom < spaceTop
		? {
				alignment: isTop
					? flipHorizontal(preferredAlignment)
					: preferredAlignment,
				rect: bottomRect,
		  }
		: {
				alignment: isTop
					? preferredAlignment
					: flipHorizontal(preferredAlignment),
				rect: topRect,
		  };
};

type RectForAlignmentFactoryFn = (
	triggerRect: Readonly<Rect>,
	triggerOffset: number,
	containerDimensions: Readonly<Dimensions>,
) => (alignment: EAlignment) => Rect;
const getRectForAlignmentFactory: RectForAlignmentFactoryFn = (
	triggerRect,
	triggerOffset,
	containerDimensions,
) => alignment =>
	getRectForAlignment(
		alignment,
		triggerRect,
		containerDimensions,
		triggerOffset,
	);

/**
 * @description
 * Given an alignment, we'll return a "base" rect, for that alignment relative to the trigger's rect.
 */
const getRectForAlignment = (
	alignment: EAlignment,
	triggerRect: Readonly<Rect>,
	containerDimensions: Readonly<Dimensions>,
	triggerOffset: number,
): Rect => {
	const alignedTopY =
		triggerRect.top - containerDimensions.height - triggerOffset;
	const alignedBottomY = triggerRect.bottom + triggerOffset;
	const centerX =
		triggerRect.left +
		triggerRect.width / 2 -
		containerDimensions.width / 2;
	const centerY =
		triggerRect.top +
		triggerRect.height / 2 -
		containerDimensions.height / 2;

	switch (alignment) {
		case EAlignment.TOP_LEFT:
			return makeRect(containerDimensions, {
				top: alignedTopY,
				left: triggerRect.left,
			});
		case EAlignment.TOP:
			return makeRect(containerDimensions, {
				top: alignedTopY,
				left: centerX,
			});
		case EAlignment.TOP_RIGHT:
			return makeRect(containerDimensions, {
				top: alignedTopY,
				left: triggerRect.right - containerDimensions.width,
			});
		case EAlignment.RIGHT:
			return makeRect(containerDimensions, {
				top: centerY,
				left: triggerRect.right + triggerOffset,
			});
		case EAlignment.BOTTOM_RIGHT:
			return makeRect(containerDimensions, {
				top: alignedBottomY,
				left: triggerRect.right - containerDimensions.width,
			});
		case EAlignment.BOTTOM:
			return makeRect(containerDimensions, {
				top: alignedBottomY,
				left: centerX,
			});
		default:
		case EAlignment.BOTTOM_LEFT:
			return makeRect(containerDimensions, {
				top: alignedBottomY,
				left: triggerRect.left,
			});
		case EAlignment.LEFT:
			return makeRect(containerDimensions, {
				top: centerY,
				left:
					triggerRect.left -
					containerDimensions.width -
					triggerOffset,
			});
	}
};

/**
 * @description
 * A helper function, that will "create" a Rect compatible object we use to pass around.
 */
export const makeRect = (
	dimensions: Dimensions,
	position: Position,
): Readonly<Rect> => ({
	...dimensions,
	top: Math.ceil(position.top),
	left: Math.ceil(position.left),
	right: Math.ceil(position.left) + dimensions.width,
	bottom: Math.ceil(position.top) + dimensions.height,
});

const getViewportDimensions = (): Dimensions => ({
	height: document.documentElement.clientHeight,
	width: document.documentElement.clientWidth,
});

const flipHorizontal = (alignment: EAlignment): EAlignment => {
	switch (alignment) {
		case EAlignment.TOP_LEFT:
			return EAlignment.BOTTOM_LEFT;
		case EAlignment.TOP:
		default:
			return EAlignment.BOTTOM;
		case EAlignment.TOP_RIGHT:
			return EAlignment.BOTTOM_RIGHT;
		case EAlignment.BOTTOM_LEFT:
			return EAlignment.TOP_LEFT;
		case EAlignment.BOTTOM:
			return EAlignment.TOP;
		case EAlignment.BOTTOM_RIGHT:
			return EAlignment.TOP_RIGHT;
	}
};

const isAlignedOnTop = (alignment: EAlignment): boolean => {
	switch (alignment) {
		case EAlignment.TOP_LEFT:
		case EAlignment.TOP:
		case EAlignment.TOP_RIGHT:
			return true;
		default:
			return false;
	}
};

const isAlignedHorizontal = (alignment: EAlignment): boolean => {
	switch (alignment) {
		case EAlignment.LEFT:
		case EAlignment.RIGHT:
			return true;
		default:
			return false;
	}
};

// Thanks to the wonderful guys over at Segment, and Evergreen for the technical inspiration for this file. <3
