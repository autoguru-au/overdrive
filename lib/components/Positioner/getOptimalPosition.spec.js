import { getOptimalPosition, makeRect } from './getOptimalPosition';
import { EAlignment } from './alignment';

const getTriggerRect = () =>
	makeRect(
		{
			width: 100,
			height: 50,
		},
		{
			left: 100,
			top: 100,
		}
	);

const viewport = overrides => ({ width: 1000, height: 1000, ...overrides });

describe('getOptimalPosition', () => {
	it('should not throw', () => {
		expect(() =>
			getOptimalPosition(EAlignment.BOTTOM_RIGHT, getTriggerRect(), {
				width: 100,
				height: 100,
			})
		).not.toThrow();
	});

	it('should match snapshot for all alignments', () => {
		const alignments = Object.entries(EAlignment).map(([, alignment]) =>
			getOptimalPosition(
				alignment,
				getTriggerRect(),
				{
					width: 5,
					height: 5,
				},
				void 0,
				void 0,
				viewport()
			)
		);

		expect(alignments).toMatchSnapshot();
	});

	it('should lock to the viewport when trigger is "off screen"', () => {
		const containerHeight = 100;

		const { alignment, rect } = getOptimalPosition(
			EAlignment.BOTTOM,
			makeRect({ width: 100, height: 50 }, { top: 1200, left: 100 }),
			{
				width: 50,
				height: containerHeight,
			},
			0,
			0,
			viewport({ height: 500 })
		);

		expect(alignment).toEqual(EAlignment.TOP);
		expect(rect.bottom - rect.top).toEqual(100);
		expect(rect.top).toBeLessThan(500);
		expect(rect.top).toEqual(500 - containerHeight);
	});

	describe('when #getFittedPosition', () => {
		it('should have flipped to left, for a right', () => {
			const { alignment } = getOptimalPosition(
				EAlignment.RIGHT,
				getTriggerRect(),
				{
					width: 50,
					height: 100,
				},
				void 0,
				void 0,
				viewport({ width: 225 })
			);

			expect(alignment).toEqual(EAlignment.LEFT);
		});

		it('should have flipped to right, for a left', () => {
			const { alignment } = getOptimalPosition(
				EAlignment.LEFT,
				makeRect(
					{
						width: 100,
						height: 50,
					},
					{
						top: 100,
						left: 50,
					}
				),
				{
					width: 100,
					height: 100,
				},
				void 0,
				void 0,
				viewport()
			);

			expect(alignment).toEqual(EAlignment.RIGHT);
		});

		it('should have flipped to the bottom, for a top', () => {
			const { alignment } = getOptimalPosition(
				EAlignment.TOP,
				makeRect(
					{
						width: 100,
						height: 50,
					},
					{ top: 50, left: 100 }
				),
				{
					width: 100,
					height: 50,
				},
				void 0,
				void 0,
				viewport({ height: 175 })
			);

			expect(alignment).toEqual(EAlignment.BOTTOM);
		});

		it('should be placed right when there is more room', () => {
			const { alignment } = getOptimalPosition(
				EAlignment.LEFT,
				makeRect(
					{
						width: 10,
						height: 50,
					},
					{ top: 50, left: 5 }
				),
				{
					width: 50,
					height: 50,
				},
				0,
				0,
				viewport({ width: 50 })
			);

			expect(alignment).toEqual(EAlignment.RIGHT);
		});
	});

	describe('when #handleOverflow', () => {
		it('should shift to the right, when on the left and no room', () => {
			const { alignment, rect } = getOptimalPosition(
				EAlignment.RIGHT,
				makeRect(
					{
						width: 20,
						height: 50,
					},
					{ top: 50, left: 150 }
				),
				{
					width: 200,
					height: 50,
				},
				0,
				0,
				viewport({ width: 200 })
			);

			expect(alignment).toEqual(EAlignment.LEFT);
			expect(rect.left).toEqual(0);
		});
	});
});
