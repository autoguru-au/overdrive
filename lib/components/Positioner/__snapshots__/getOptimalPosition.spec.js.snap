// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOptimalPosition should match snapshot for all alignments 1`] = `
Array [
  Object {
    "alignment": "top",
    "rect": Object {
      "bottom": 88,
      "height": 5,
      "left": 148,
      "right": 153,
      "top": 83,
      "width": 5,
    },
  },
  Object {
    "alignment": "top_left",
    "rect": Object {
      "bottom": 88,
      "height": 5,
      "left": 100,
      "right": 105,
      "top": 83,
      "width": 5,
    },
  },
  Object {
    "alignment": "top_right",
    "rect": Object {
      "bottom": 88,
      "height": 5,
      "left": 195,
      "right": 200,
      "top": 83,
      "width": 5,
    },
  },
  Object {
    "alignment": "bottom",
    "rect": Object {
      "bottom": 167,
      "height": 5,
      "left": 148,
      "right": 153,
      "top": 162,
      "width": 5,
    },
  },
  Object {
    "alignment": "bottom_left",
    "rect": Object {
      "bottom": 167,
      "height": 5,
      "left": 100,
      "right": 105,
      "top": 162,
      "width": 5,
    },
  },
  Object {
    "alignment": "bottom_right",
    "rect": Object {
      "bottom": 167,
      "height": 5,
      "left": 195,
      "right": 200,
      "top": 162,
      "width": 5,
    },
  },
  Object {
    "alignment": "left",
    "rect": Object {
      "bottom": 128,
      "height": 5,
      "left": 83,
      "right": 88,
      "top": 123,
      "width": 5,
    },
  },
  Object {
    "alignment": "right",
    "rect": Object {
      "bottom": 128,
      "height": 5,
      "left": 212,
      "right": 217,
      "top": 123,
      "width": 5,
    },
  },
]
`;
